### ISSUE TITLE
Trait bounds on struct type parameters aren't always enforced (depending on the order of impls)

### ISSUE THREAD

**Issue by SSheldon** [rust-lang/rust#21837]
The following compiles without issue, even though the impl of `Trait2` is missing a trait bound that `T` implements `Bound`:
``` rust
pub trait Bound { }
pub struct Foo<T: Bound>;

pub trait Trait1 { }
impl<T: Bound> Trait1 for Foo<T> { }

pub trait Trait2 { }
impl<T> Trait2 for Foo<T> { }

fn main() { }
```

However, if you switch the order of the impls...
``` rust
pub trait Bound { }
pub struct Foo<T: Bound>;

pub trait Trait2 { }
impl<T> Trait2 for Foo<T> { }

pub trait Trait1 { }
impl<T: Bound> Trait1 for Foo<T> { }

fn main() { }
```

You suddenly start getting this error:
```
error: the trait `Bound` is not implemented for the type `T` [E0277]
impl<T> Trait2 for Foo<T> { }
^~~~~~~~~~~~~~~~~~~~~~~~~~~~~
error: aborting due to previous error
```

I did not expect the ordering of the impls to matter here. I'm using `rustc 1.0.0-nightly (1d00c545e 2015-01-30 19:56:34 +0000)`.

**Comment by pnkfelix** [comment]
P-high, not 1.0.  (But definitely a weird bug that we would like to fix.)

**Comment by nikomatsakis** [comment]
cc me

**Comment by SSheldon** [comment]
@nikomatsakis, @pnkfelix: should this be P-backcompat-lang since the compiler accepts code that it should not?

**Comment by pnkfelix** [comment]
@SSheldon I think we decided this is one of those edge cases where the amount of breakage (that fixing this is expected to incur) was deemed acceptable for fixing post 1.0.

At the time this was last triaged, there was a general rule that if something was tagged P-backcompat-lang, then it must also be set for the 1.0 milestone (at the latest).

However, since then during recent triage meetings, we have been more willing to tag things as P-backcompat-lang even if they were not attached to the 1.0 milestone.

So, with that in mind:

triage: P-backcompat-lang, not P-high

**Comment by arielb1** [comment]
This causes an RFC1214 warning in 1.4+.

**Comment by nikomatsakis** [comment]
triage: P-medium

Looks like this is already fixed by RFC 1214, as @arielb1 reports.

**Comment by nikomatsakis** [comment]
Tagging as E-needstest, since we should be able to add a test with `WARN` messages (and perhaps a `#[rustc_error]` annotation on `main`)

**Comment by nikomatsakis** [comment]
Tagging as mentor just because I can give advice on how to craft a suitable test (`nmatsakis` on IRC)

**Comment by bltavares** [comment]
Triaging:

Both examples causes error on nightly `rustc 1.11.0-nightly (5c2a5d449 2016-06-11)`, but this seems to be the expected behaviour.

```
<anon>:2:16: 2:17 error: parameter `T` is never used [E0392]
<anon>:2 pub struct Foo<T: Bound>;
                        ^
<anon>:2:16: 2:17 help: see the detailed explanation for E0392
<anon>:2:16: 2:17 help: consider removing `T` or using a marker such as `core::marker::PhantomData`
<anon>:8:9: 8:15 error: the trait `Bound` is not implemented for the type `T` [E0277]
<anon>:8 impl<T> Trait2 for Foo<T> { }
                 ^~~~~~
<anon>:8:9: 8:15 help: see the detailed explanation for E0277
<anon>:8:9: 8:15 note: required by `Foo`
error: aborting due to 2 previous errors
```

@nikomatsakis I would like to help with the test if you don't mind giving initial directions.

**Comment by SSheldon** [comment]
@bltavares, that repro case was from pre-1.0, here's the updated repro case:
``` rust
pub trait Bound { }
pub struct Foo<T: Bound>(T);

pub trait Trait1 { }
impl<T: Bound> Trait1 for Foo<T> { }

pub trait Trait2 { }
impl<T> Trait2 for Foo<T> { }

fn main() { }
```

That code will compile with `rustc 1.0.0 (a59de37e9 2015-05-13)`, but fails to compile with `rustc 1.9.0 (e4e8b6668 2016-05-18)`.

**Comment by brson** [comment]
@arielb1 This seems to have been fixed by RFC 1214. Can it be closed?

**Comment by pnkfelix** [comment]
@brson I think it needs a test. I probably should not be assigned to it since I don't plan to spend time making a test.

**Comment by martinhath** [comment]
So all there is to do is add a test which checks that Ssheldons code does not compile?

**Comment by nikomatsakis** [comment]
@martinhath yes :)

**Comment by SSheldon** [comment]
Fixed now with a test was merged in #36941
